File: src/java/IndentWriter.java
package scala.virtualization.lms.epfl.test7;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.Writer;

class IndentWriter extends Writer {
    private Writer out;
    private int indent;
    private final char INDENT_INC = '{';
    private final char INDENT_DEC = '}';
    private final char INDENT_DO = '\n';
    private final String INDENT_STR = "  ";
    private boolean doIndent;

    public IndentWriter(Writer out) {
        this.out = out;
        this.indent = 0;
        this.doIndent = false;
    }

    public void write(char[] cbuf, int off, int len) throws IOException {
        // No knowledge about strings or anything else...
        for (int i=off; i<len; i++) {
            if (cbuf[i] == INDENT_DEC)
                indent--;

            if (doIndent)
                for (int j=0; j<indent; j++)
                    out.write(INDENT_STR);
            doIndent = false;

            out.write(cbuf[i]);

            if (cbuf[i] == INDENT_INC)
                indent++;
            else if (cbuf[i] == INDENT_DO)
                doIndent = true;
        }
    }

    public void close() throws IOException {
        out.close();
    }

    public void flush() throws IOException {
        out.flush();
    }

    public static PrintWriter getIndentPrintWriter(Writer out) {
        return new PrintWriter(new IndentWriter(out));
    }
}

File: src/scala/virtualization/lms/util/IndentWriter.scala
package scala.virtualization.lms.util

import java.io.IOException
import java.io.Writer

class IndentWriter(out: Writer) extends Writer {
    private var indent: Int = 0
    private val INDENT_INC: Char = '{'
    private val INDENT_DEC: Char = '}'
    private val INDENT_DO: Char = '\n'
    private val INDENT_STR: String = "  "
    private var doIndent: Boolean = false

    override def write(cbuf: Array[Char], off: Int, len:Int) {
      var i = off;
      var j = 0;

      while (i < off + len) {
        if (cbuf(i) == INDENT_DEC)
            indent -= 1;

        if (doIndent) {
          j = 0
          while(j < indent) {
            out.write(INDENT_STR)
            j += 1
          }
          doIndent = false
        }

        out.write(cbuf(i));

        if (cbuf(i) == INDENT_INC)
            indent += 1;
        else if (cbuf(i) == INDENT_DO)
            doIndent = true;

        i += 1
      }
    }

    def close() = out.close()
    def flush() = out.flush()
}
